### ISSUE TITLE
Optimizer state gets automatically restored when loading weights from checkpoint and doesnt change when you compile the model

### ISSUE THREAD

**Issue by Shubham3101** [tensorflow/tensorflow#33842]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: None
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): TF 2.0.0(stable)
- Python version: 3.6
- Bazel version (if compiling from source):None
- GCC/Compiler version (if compiling from source): None

**Describe the current behavior**
Optimizer state is getting restored while loading weights from model checkpoint(tf.keras.callbacks.ModelCheckpoint). Later when you compile model with different learning rate(different from the restored optimizer state), it doesn't gets updated.  

**Describe the expected behavior**
Compiling the model should update the optimizer state

**Code to reproduce the issue**

```python
import tensorflow as tf
from tensorflow.python.keras import backend as K
import numpy as np

def get_data():
    images = np.zeros((64,224))
    labels = np.zeros((64,5))
    return images,labels

def create_model():
    input_layer = tf.keras.layers.Input(name='Image_input', shape=(224), dtype='float32')
    model = tf.keras.layers.Dense(5)(input_layer)
    model = tf.keras.layers.Activation('softmax', name = "output-softmax")(model)
    model = tf.keras.models.Model(inputs=input_layer, outputs=[model])
    return model

os.makedirs("checkpoints/")
checkpoint = tf.keras.callbacks.ModelCheckpoint("checkpoints/", monitor='loss',
                                                save_weights_only=True, save_freq='epoch')
reduce_lr = tf.keras.callbacks.ReduceLROnPlateau(monitor='loss', factor=0.1, patience=1, verbose=1)
early_stopping = tf.keras.callbacks.EarlyStopping(monitor='loss', min_delta=0, patience=5, verbose=1)

model = create_model()
data = get_data()
​
optimizer = tf.keras.optimizers.Adadelta(learning_rate=1.0)
model.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])
​
print ("learning_rate",float(K.get_value(model.optimizer.lr)))
                                                       
model.fit(data[0], data[1], batch_size=16, epochs=10,callbacks=[checkpoint,reduce_lr,early_stopping])

print ("**learning_rate_1**",float(K.get_value(model.optimizer.lr)))`

model = create_model()

model.load_weights('checkpoints/')

optimizer = tf.keras.optimizers.Adadelta(learning_rate=0.1)
model.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])

print ("learning_rate",float(K.get_value(model.optimizer.lr)))`
```

final learning rate should be **0.1** but it is equal to **learning_rate_1** .

**Comment by ravikyram** [comment]
I have tried on colab with TF version 2.0  and was able to reproduce the issue.Please, find the gist[ here](https://colab.sandbox.google.com/gist/ravikyram/2a4efeeee5fb43761848ea1c429c3e36/untitled320.ipynb). Thanks!

**Comment by bono1567** [comment]
The "K.get_value" seems to be making the change.

**Comment by Shubham3101** [comment]
I don't think so, because if you compile the again (2nd time) , the optimizer state gets updated

**Comment by ravikyram** [comment]
I have tried on colab with TF version 2.2-rc4 and was able to reproduce the issue.Please, find the gist [here](https://colab.sandbox.google.com/gist/ravikyram/010fce8a813d407a528dd37847a19a24/untitled855.ipynb). Thanks!

**Comment by amahendrakar** [comment]
Was able to reproduce the issue with [TF v2.2](https://colab.research.google.com/gist/amahendrakar/897a7b0d1d9c928b89f6cab443615645/33842.ipynb) and [TF-nightly](https://colab.research.google.com/gist/amahendrakar/2d40870dabfc9294880ba74d367d02f2/33842-tf-nightly.ipynb). Please find the attached gist. Thanks!

**Comment by bhack** [comment]
I suppose it is something related to the deferred restore of optimizer value. Cause in `h5` is working:
```
import tensorflow as tf
from tensorflow.python.keras import backend as K
import numpy as np
import os
def get_data():
    images = np.zeros((64,224))
    labels = np.zeros((64,5))
    return images,labels

def create_model():
    input_layer = tf.keras.layers.Input(name='Image_input', shape=(224), dtype='float32')
    model = tf.keras.layers.Dense(5)(input_layer)
    model = tf.keras.layers.Activation('softmax', name = "output-softmax")(model)
    model = tf.keras.models.Model(inputs=input_layer, outputs=[model])
    return model

#os.makedirs("checkpoints/")


model = create_model()
data = get_data()

optimizer = tf.keras.optimizers.Adadelta(learning_rate=0.1)
model.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])
print ("learning_rate",float(K.get_value(model.optimizer.lr)))
                                                       
model.fit(data[0], data[1], batch_size=16, epochs=10)
model.save_weights('/checkpoints_test/test.h5')
print ("**learning_rate_1**",float(K.get_value(model.optimizer.lr)))
model = create_model()

model.load_weights('/checkpoints_test/test.h5')
optimizer = tf.keras.optimizers.SGD(learning_rate=1)
model.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])

print ("learning_rate",float(K.get_value(optimizer.lr)))
```

**Comment by sachinprasadhs** [comment]
Was able to reproduce the issue in Tensorflow version 2.5, please find the gist [here](https://colab.research.google.com/gist/googly789/3f974e5f5efa0eaa39902e6a013d29f1/untitled40.ipynb).

**Comment by sushreebarsa** [comment]
@Shubham3101 Was able to reproduce the issue on colab using TF v2.7.0, please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/41ec5087955cb5742591f3f9fd7b8375/untitled40.ipynb#scrollTo=TvcImRC67wGo).Thank you!

**Comment by Shubham3101** [comment]
@sushreebarsa , I think it’s the default behaviour to load the optimiser state with weights. 
once it’s loaded, you can compile it and start the training from the restored state. 

I am closing the issue for now, feel free to open it if required
